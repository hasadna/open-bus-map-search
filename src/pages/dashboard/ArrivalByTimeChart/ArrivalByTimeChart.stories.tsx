import { Meta, StoryObj } from '@storybook/react'
import ArrivalByTimeChart from './ArrivalByTimeChart'
import testBusData from './testdata/data.json'

const meta: Meta<typeof ArrivalByTimeChart> = {
  title: 'Components/Dashboard/ArrivalByTimeChart',
  component: ArrivalByTimeChart,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof ArrivalByTimeChart>

export const Default: Story = {
  args: {
    data: testBusData,
    operatorId: testBusData[0].id,
  },
}
