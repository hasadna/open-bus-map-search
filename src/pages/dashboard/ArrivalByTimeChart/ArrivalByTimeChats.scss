.chart{
  scroll-snap-type: y mandatory;
  & > div{
    scroll-snap-align: start;
  }
}