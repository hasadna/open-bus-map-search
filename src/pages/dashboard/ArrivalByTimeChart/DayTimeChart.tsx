import React, { FC, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { Skeleton, Radio, RadioChangeEvent } from 'antd'
import ArrivalByTimeChart from './ArrivalByTimeChart'
import { GroupByRes, useGroupBy } from 'src/api/groupByService'
import { Moment } from 'moment/moment'
import Widget from 'src/shared/Widget'

const convertToGraphCompatibleStruct = (arr: GroupByRes[]) => {
  return arr.map((item: GroupByRes) => ({
    id: item.operator_ref?.agency_id || 'Unknown',
    name: item.operator_ref?.agency_name || 'Unknown',
    current: item.total_actual_rides,
    max: item.total_planned_rides,
    percent: (item.total_actual_rides / item.total_planned_rides) * 100,
    gtfs_route_date: item.gtfs_route_date,
    gtfs_route_hour: item.gtfs_route_hour,
  }))
}

interface DayTimeChartProps {
  startDate: Moment
  endDate: Moment
  operatorId: string
}

const DayTimeChart: FC<DayTimeChartProps> = ({ startDate, endDate, operatorId }) => {
  const { t } = useTranslation()
  const [groupByHour, setGroupByHour] = React.useState<boolean>(false)

  const [data, loadingGraph] = useGroupBy({
    dateTo: endDate,
    dateFrom: startDate,
    groupBy: groupByHour ? 'operator_ref,gtfs_route_hour' : 'operator_ref,gtfs_route_date',
  })

  const graphData = useMemo(
    () => convertToGraphCompatibleStruct(data),
    [endDate, groupByHour, startDate, data.length],
  )

  return (
    <Widget>
      <h2 className="title">
        {groupByHour ? t('dashboard_page_graph_title_hour') : t('dashboard_page_graph_title_day')}
      </h2>
      <Radio.Group
        style={{ marginBottom: '10px' }}
        onChange={(e: RadioChangeEvent) => setGroupByHour(e.target.value === 'byHour')}
        defaultValue="byDay">
        <Radio.Button value="byDay">{t('group_by_day_tooltip_content')}</Radio.Button>
        <Radio.Button value="byHour">{t('group_by_hour_tooltip_content')}</Radio.Button>
      </Radio.Group>
      {loadingGraph ? (
        <Skeleton active />
      ) : (
        <ArrivalByTimeChart data={graphData} operatorId={operatorId} />
      )}
    </Widget>
  )
}

export default DayTimeChart
