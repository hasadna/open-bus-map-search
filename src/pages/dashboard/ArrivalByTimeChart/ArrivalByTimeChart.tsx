import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from 'recharts'

import './ArrivalByTimeChats.scss'
import { useMemo } from 'react'
import moment from 'moment-timezone'

export const arrayGroup = function <T>(array: T[], f: (item: T) => string) {
  const groups: Record<string, T[]> = {}
  array.forEach(function (o) {
    const group = f(o)
    groups[group] = groups[group] || []
    groups[group].push(o)
  })
  return Object.keys(groups).map(function (group) {
    return groups[group]
  })
}

export default function ArrivalByTimeChart({
  data,
  operatorId,
}: {
  data: {
    id: string
    name: string
    current: number
    max: number
    percent: number
    gtfs_route_date: string
    gtfs_route_hour: string
  }[]
  operatorId: string
}) {
  data = useMemo(
    () => data.filter((item) => !operatorId || item.id === operatorId),
    [data, operatorId],
  )
  return (
    <div className="chart">
      {arrayGroup(data, (item) => item.id)
        .filter((group) => group.length > 1 && group.reduce((a, b) => a + b.current, 0) > 1)
        .map((group) => (
          <div key={group[0].name}>
            <h3 className="title">{group[0].name}</h3>
            <ResponsiveContainer debounce={1000} width="100%" height={300}>
              <LineChart
                data={group
                  .sort((a, b) => (a.gtfs_route_date < b.gtfs_route_date ? 1 : -1))
                  .sort((a, b) => (a.gtfs_route_hour > b.gtfs_route_hour ? 1 : -1))}
                margin={{
                  top: 5,
                  right: 30,
                  left: 20,
                  bottom: 5,
                }}>
                <CartesianGrid strokeDasharray="3 3" />
                <XAxis dataKey="gtfs_route_date" />
                <YAxis />
                <Tooltip
                  content={({ payload }) =>
                    payload?.[0] && (
                      <>
                        <ul>
                          <li>
                            <span className="label">זמן: </span>
                            <span className="value">
                              {payload![0].payload.gtfs_route_date
                                ? moment
                                    .utc(payload![0].payload.gtfs_route_date)
                                    .format('יום ddd, L')
                                : moment(payload![0].payload.gtfs_route_hour).format(
                                    'יום dddd, LT',
                                  )}
                            </span>
                          </li>
                          <li>
                            <span className="label">ביצוע: </span>
                            <span className="value">{payload![0].payload.current}</span>
                          </li>
                          <li>
                            <span className="label">תכנון: </span>
                            <span className="value">{payload![0].payload.max}</span>
                          </li>
                          <li>
                            <span className="label">דיוק: </span>
                            <span className="value">
                              {(
                                (payload![0].payload.current / payload![0].payload.max) *
                                100
                              ).toFixed(2)}
                              %
                            </span>
                          </li>
                        </ul>
                      </>
                    )
                  }
                />
                <Legend />
                <Line type="monotone" dataKey="percent" stroke="#8884d8" activeDot={{ r: 8 }} />
              </LineChart>
            </ResponsiveContainer>
          </div>
        ))}
    </div>
  )
}
