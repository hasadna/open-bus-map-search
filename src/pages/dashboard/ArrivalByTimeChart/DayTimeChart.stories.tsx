import type { Meta, StoryObj } from '@storybook/react'
import moment from 'moment'
import DayTimeChart from './DayTimeChart'

const meta: Meta<typeof DayTimeChart> = {
  component: DayTimeChart,
  title: 'Pages/Dashboard/ArrivalByTimeChart',
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
  argTypes: {
    startDate: {
      control: 'date',
      description: 'The start date of the chart.',
      table: {
        type: {
          summary: 'Moment',
        },
      },
    },
    endDate: {
      control: 'date',
      description: 'The end date of the chart.',
      table: {
        type: {
          summary: 'Moment',
        },
      },
    },
    operatorId: {
      control: 'text',
      description: 'The operator id of the chart.',
    },
  },
  render: (args) => (
    <DayTimeChart
      startDate={moment(args.startDate)}
      endDate={moment(args.endDate)}
      operatorId={args.operatorId}
    />
  ),
}

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    startDate: moment(),
    endDate: moment().add(-7, 'day'),
    operatorId: '3',
  },
}
