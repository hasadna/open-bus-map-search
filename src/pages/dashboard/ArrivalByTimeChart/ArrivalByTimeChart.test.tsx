import { type RenderResult, render, screen } from '@testing-library/react'
import ArrivalByTimeChart from './ArrivalByTimeChart'
import type { ReactElement } from 'react'
import testBusData from './testdata/data.json'

jest.mock('recharts', () => {
  const original: typeof import('recharts') = jest.requireActual('recharts')
  return {
    __esModule: true,
    ...original,
    ResponsiveContainer: jest
      .fn()
      .mockImplementation(({ children }: { children: ReactElement }) => (
        <original.ResponsiveContainer height={300} aspect={1}>
          {children}
        </original.ResponsiveContainer>
      )),
  }
})

describe('ArrivalByTimeChart', () => {
  let renderedComponent: RenderResult
  beforeEach(() => {
    renderedComponent = render(
      <ArrivalByTimeChart data={testBusData} operatorId={testBusData[0].id} />,
    )
  })

  test('renders without crashing', () => {
    expect(screen.getByText(testBusData[0].name)).toBeInTheDocument()
  })

  test('tooltip wrapper exists', () => {
    expect(renderedComponent.container.querySelector('.recharts-tooltip-wrapper')).not.toBeNull()
  })

  test('tooltips are not visible by default', () => {
    expect(renderedComponent.container.querySelector('.recharts-tooltip-wrapper')).not.toBeVisible()
  })
  test('filters operators correctly', () => {
    expect(screen.queryByText('notegged')).not.toBeInTheDocument()
    expect(screen.queryByText('Unknown')).not.toBeInTheDocument()
  })
})
