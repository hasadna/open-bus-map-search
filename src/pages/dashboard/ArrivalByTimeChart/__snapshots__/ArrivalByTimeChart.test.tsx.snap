// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArrivalByTimeChart > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="chart"
  >
    <div>
      <h3
        class="title"
      >
        אגד
      </h3>
      <div
        class="recharts-responsive-container"
        style="width: 800px; height: 800px; min-width: 0;"
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 800px; height: 800px;"
        >
          <svg
            class="recharts-surface"
            height="800"
            viewBox="0 0 800 800"
            width="800"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="760"
                  width="690"
                  x="80"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-cartesian-grid"
            >
              <g
                class="recharts-cartesian-grid-horizontal"
              >
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="80"
                  x2="770"
                  y="5"
                  y1="765"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="80"
                  x2="770"
                  y="5"
                  y1="575"
                  y2="575"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="80"
                  x2="770"
                  y="5"
                  y1="385"
                  y2="385"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="80"
                  x2="770"
                  y="5"
                  y1="195"
                  y2="195"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="80"
                  x2="770"
                  y="5"
                  y1="5"
                  y2="5"
                />
              </g>
              <g
                class="recharts-cartesian-grid-vertical"
              >
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="80"
                  x2="80"
                  y="5"
                  y1="5"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="195"
                  x2="195"
                  y="5"
                  y1="5"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="310"
                  x2="310"
                  y="5"
                  y1="5"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="425"
                  x2="425"
                  y="5"
                  y1="5"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="540"
                  x2="540"
                  y="5"
                  y1="5"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="655"
                  x2="655"
                  y="5"
                  y1="5"
                  y2="765"
                />
                <line
                  fill="none"
                  height="760"
                  stroke="#ccc"
                  stroke-dasharray="3 3"
                  width="690"
                  x="80"
                  x1="770"
                  x2="770"
                  y="5"
                  y1="5"
                  y2="765"
                />
              </g>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="690"
                x="80"
                x1="80"
                x2="770"
                y="765"
                y1="765"
                y2="765"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              >
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="80"
                    x2="80"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="80"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="80"
                    >
                      2024-01-22
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="195"
                    x2="195"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="195"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="195"
                    >
                      2024-01-23
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="310"
                    x2="310"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="310"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="310"
                    >
                      2024-01-24
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="425"
                    x2="425"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="425"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="425"
                    >
                      2024-01-25
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="540"
                    x2="540"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="540"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="540"
                    >
                      2024-01-26
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="655"
                    x2="655"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="655"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="655"
                    >
                      2024-01-27
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="690"
                    x="80"
                    x1="770"
                    x2="770"
                    y="765"
                    y1="771"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="690"
                    x="770"
                    y="773"
                  >
                    <tspan
                      dy="0.71em"
                      x="770"
                    >
                      2024-01-28
                    </tspan>
                  </text>
                </g>
              </g>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="760"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="80"
                x2="80"
                y="5"
                y1="5"
                y2="765"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              >
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="760"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="20"
                    x1="74"
                    x2="80"
                    y="5"
                    y1="765"
                    y2="765"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="760"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="72"
                    y="765"
                  >
                    <tspan
                      dy="0.355em"
                      x="72"
                    >
                      0
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="760"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="20"
                    x1="74"
                    x2="80"
                    y="5"
                    y1="575"
                    y2="575"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="760"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="72"
                    y="575"
                  >
                    <tspan
                      dy="0.355em"
                      x="72"
                    >
                      25
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="760"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="20"
                    x1="74"
                    x2="80"
                    y="5"
                    y1="385"
                    y2="385"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="760"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="72"
                    y="385"
                  >
                    <tspan
                      dy="0.355em"
                      x="72"
                    >
                      50
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="760"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="20"
                    x1="74"
                    x2="80"
                    y="5"
                    y1="195"
                    y2="195"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="760"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="72"
                    y="195"
                  >
                    <tspan
                      dy="0.355em"
                      x="72"
                    >
                      75
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="760"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="20"
                    x1="74"
                    x2="80"
                    y="5"
                    y1="5"
                    y2="5"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="760"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="72"
                    y="5"
                  >
                    <tspan
                      dy="0.355em"
                      x="72"
                    >
                      100
                    </tspan>
                  </text>
                </g>
              </g>
            </g>
            <g
              class="recharts-layer recharts-line"
            >
              <path
                class="recharts-curve recharts-line-curve"
                d="M80,13.312204822271903C118.33333333333334,12.917232158102403,156.66666666666666,12.522259493932903,195,12.522259493932903C233.33333333333334,12.522259493932903,271.6666666666667,12.761739192205733,310,13.24069858875139C348.3333333333333,13.719657985297049,386.6666666666667,14.202265344915102,425,15.468147905222622C463.3333333333333,16.734030465530143,501.6666666666667,18.72561280436551,540,20.835993950596517C578.3333333333334,22.946375096827524,616.6666666666666,28.13043478260865,655,28.13043478260865C693.3333333333334,28.13043478260865,731.6666666666666,25.520540703957877,770,22.910646625307105"
                fill="none"
                height="760"
                stroke="#8884d8"
                stroke-dasharray="0px 0px"
                stroke-width="1"
                width="690"
              />
            </g>
          </svg>
          <div
            class="recharts-legend-wrapper"
            style="position: absolute; width: 750px; height: auto; left: 20px; bottom: 5px;"
          >
            <ul
              class="recharts-default-legend"
              style="padding: 0px; margin: 0px; text-align: center;"
            >
              <li
                class="recharts-legend-item legend-item-0"
                style="display: inline-block; margin-right: 10px;"
              >
                <svg
                  class="recharts-surface"
                  height="14"
                  style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                  viewBox="0 0 32 32"
                  width="14"
                >
                  <title />
                  <desc />
                  <path
                    class="recharts-legend-icon"
                    d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                    fill="none"
                    stroke="#8884d8"
                    stroke-width="4"
                  />
                </svg>
                <span
                  class="recharts-legend-item-text"
                  style="color: rgb(136, 132, 216);"
                >
                  percent
                </span>
              </li>
            </ul>
          </div>
          <div
            class="recharts-tooltip-wrapper"
            role="dialog"
            style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
